clippy_lints/src/hash_borrow_str_semantics.rs

```diff
@@ -0,0 +1,103 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::ty::implements_trait;
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::{Item, ItemKind, Path, TraitRef};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::Ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// This lint is concerned with the semantics of Borrow and Hash for a
+    /// type that implements all three of Hash, Borrow<str> and Borrow<[u8]>
+    /// as it is impossible to satisfy the semantics of Borrow and Hash for
+    /// both Borrow<str> and Borrow<[u8]>.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// When providing implementations for Borrow<T>, one should consider whether the different
+    /// implementations should act as facets or representations of the underlying type. Generic code
+    /// typically uses Borrow<T> when it relies on the identical behavior of these additional trait
+    /// implementations. These traits will likely appear as additional trait bounds.
+    ///
+    /// In particular Eq, Ord and Hash must be equivalent for borrowed and owned values:
+    /// `x.borrow() == y.borrow()` should give the same result as `x == y`.
+    /// It follows then that the following equivalence must hold:
+    /// `hash(x) == hash((x as Borrow<[u8]>).borrow()) == hash((x as Borrow<str>).borrow())`
+    ///
+    /// Unfortunately it doesn't hold as `hash("abc") != hash("abc".as_bytes())`.
+    ///
```

Comment:
```suggestion
    /// this is because the that Hash impl for str passes an additional 0xFF byte to
    /// the hasher to avoid collisions. For example, given a tuple ("a", "bc"), and ("ab", "c"),
    /// the two tuples would have the same hash value if the 0xFF byte was not added.
    /// https://doc.rust-lang.org/stable/core/hash/trait.Hash.html#prefix-collisions
```